gaeguli/workers/gaeguli-pipeline-worker.c

```diff
@@ -14,18 +23,154 @@ intr_handler (gpointer user_data)
   return G_SOURCE_REMOVE;
 }
 
+static void
+_decodebin_pad_added (GstElement * decodebin, GstPad * pad, gpointer user_data)
+{
+  if (GST_PAD_PEER (pad) == NULL) {
+    GstElement *overlay = GST_ELEMENT (user_data);
+    g_autoptr (GstPad) sinkpad =
+        gst_element_get_static_pad (overlay, "video_sink");
+
+    gst_pad_link (pad, sinkpad);
+  }
+}
+
+static void
+_pipeline_update_vsrc_caps (GstElement * pipeline, GaeguliVideoSource source,
+    GaeguliVideoResolution resolution, guint fps)
+{
+  gint width, height, i;
+  g_autoptr (GstElement) capsfilter = NULL;
+  g_autoptr (GstCaps) caps = NULL;
+
+  if (!pipeline) {
+    return;
+  }
+
+  switch (resolution) {
+    case GAEGULI_VIDEO_RESOLUTION_640X480:
+      width = 640;
+      height = 480;
+      break;
+    case GAEGULI_VIDEO_RESOLUTION_1280X720:
+      width = 1280;
+      height = 720;
+      break;
+    case GAEGULI_VIDEO_RESOLUTION_1920X1080:
+      width = 1920;
+      height = 1080;
+      break;
+    case GAEGULI_VIDEO_RESOLUTION_3840X2160:
+      width = 3840;
+      height = 2160;
+      break;
+    default:
+      width = -1;
+      height = -1;
+      break;
+  }
+
+  caps = gst_caps_new_empty ();
+
+  for (i = 0; supported_formats[i] != NULL; i++) {
+    GstCaps *supported_caps = gst_caps_from_string (supported_formats[i]);
+    gst_caps_set_simple (supported_caps, "width", G_TYPE_INT, width, "height",
+        G_TYPE_INT, height, "framerate", GST_TYPE_FRACTION, fps, 1, NULL);
+    gst_caps_append (caps, supported_caps);
+  }
+
+  capsfilter = gst_bin_get_by_name (GST_BIN (pipeline), "caps");
+  g_object_set (capsfilter, "caps", caps, NULL);
+
+  /* Cycling vsrc through READY state prods decodebin into re-discovery of input
+   * stream format and rebuilding its decoding pipeline. This is needed when
+   * a switch is made between two resolutions that the connected camera can only
+   * produce in different output formats, e.g. a change from raw 640x480 stream
+   * to MJPEG 1920x1080.
+   *
+   * NVARGUS Camera src doesn't support this.
+   */
+  if (source != GAEGULI_VIDEO_SOURCE_NVARGUSCAMERASRC) {
+    GstState cur_state;
+
+    gst_element_get_state (pipeline, &cur_state, NULL, 0);
+    if (cur_state > GST_STATE_READY) {
+      gst_element_set_state (pipeline, GST_STATE_READY);
+      gst_element_set_state (pipeline, cur_state);
+    }
+  }
+}
+
 int
 main (int argc, char *argv[])
 {
   g_autoptr (GError) error = NULL;
   g_autoptr (GMainLoop) loop = NULL;
+  g_autoptr (GstElement) pipeline = NULL;
+  g_autoptr (GstElement) overlay = NULL;
+  g_autoptr (GstElement) decodebin = NULL;
+  g_autoptr (GstPluginFeature) feature = NULL;
+  g_autofree gchar *pipeline_str = NULL;
+  GaeguliVideoSource source = GAEGULI_VIDEO_SOURCE_UNKNOWN;
+  gboolean show_overlay = FALSE;
+  gboolean prefer_hw_decoding = FALSE;
+  GaeguliVideoResolution resolution = GAEGULI_VIDEO_RESOLUTION_UNKNOWN;
+  guint fps = 0;
+  gint ret = 0;
+
+  pipeline_str = g_strdup (argv[1]);
```

Comment:
I'd prefer not passing `pipeline_str` as a cmdline argument. Pass only the values necessary to build it and let the child call `_get_enc_string()`.